<span style="color:darkorange; font-weight:bold; font-size:17px;">The dataset which contains genres (The one sent by Anirudh) </span><span style="color:lightsalmon; font-weight:bold; font-size:13px;">book_data.csv<br></span>
<span style="color:darkred; font-weight:bold; font-size:13px;">Add the source and description here later<br></span>

<div style="background-color:cornsilk;"><span style="color:#182B49;">Importing Libraries</span></div>

In [1]:
import pandas as pd #for importing csv file
import numpy as np #for sum mathematical stuff
import matplotlib.pyplot as plt #for plotting

<div style="background-color:cornsilk;"><span style="color:#182B49;">Loading Datasets</span></div>

In [2]:
books = pd.read_csv('book_data.csv',error_bad_lines = False) 
#error_bad_lines : boolean, default True Lines with too many fields (e.g. a csv line with too many commas) will by default cause an exception to be raised, and no DataFrame will be returned. If False, then these “bad lines” will dropped from the DataFrame that is returned. (Only valid with C parser)

<div style="background-color:cornsilk;"><span style="color:#182B49;">Exploring the dataset content</span></div>

In [3]:
print("There are {} rows and {} columns in the dataset.".format(books.shape[0], books.shape[1]))

There are 54301 rows and 12 columns in the dataset.


In [4]:
books.shape #table dimensions

(54301, 12)

<div style="background-color:cornsilk;"><span style="color:#182B49;">Dataset columns</span></div>

In [5]:
#columns
np.array(books.columns)

array(['book_authors', 'book_desc', 'book_edition', 'book_format',
       'book_isbn', 'book_pages', 'book_rating', 'book_rating_count',
       'book_review_count', 'book_title', 'genres', 'image_url'],
      dtype=object)

<span style="color:#747678; font-size:14px;">Columns are <strong>'book_authors', 'book_desc', 'book_edition', 'book_format',
       'book_isbn', 'book_pages', 'book_rating', 'book_rating_count',
    'book_review_count', 'book_title', 'genres', 'image_url'</strong></span>

<span style="color:#747678; font-size:14px;">Depicting the first 5 rows of this dataset with <strong>head</strong> method.<br></span>

In [6]:
books.head()

book_authors  \
0                                    Suzanne Collins   
1                         J.K. Rowling|Mary GrandPré   
2                                         Harper Lee   
3  Jane Austen|Anna Quindlen|Mrs. Oliphant|George...   
4                                    Stephenie Meyer   

                                           book_desc  \
0  Winning will make you famous. Losing means cer...   
1  There is a door at the end of a silent corrido...   
2  The unforgettable novel of a childhood in a sl...   
3  «È cosa ormai risaputa che a uno scapolo in po...   
4  About three things I was absolutely positive.F...   

                         book_edition book_format    book_isbn book_pages  \
0                                 NaN   Hardcover  9.78044E+12  374 pages   
1                          US Edition   Paperback  9.78044E+12  870 pages   
2                    50th Anniversary   Paperback  9.78006E+12  324 pages   
3  Modern Library Classics, USA / CAN   Paperback  9.78068E+12  279 pages   
4                                 NaN   Paperback  9.78032E+12  498 pages   

   book_rating  book_rating_count  book_review_count  \
0         4.33            5519135             160706   
1         4.48            2041594              33264   
2         4.27            3745197              79450   
3         4.25            2453620              54322   
4         3.58            4281268              97991   

                                  book_title  \
0                           The Hunger Games   
1  Harry Potter and the Order of the Phoenix   
2                      To Kill a Mockingbird   
3                        Pride and Prejudice   
4                                   Twilight   

                                              genres  \
0  Young Adult|Fiction|Science Fiction|Dystopia|F...   
1                        Fantasy|Young Adult|Fiction   
2  Classics|Fiction|Historical|Historical Fiction...   
3                           Classics|Fiction|Romance   
4  Young Adult|Fantasy|Romance|Paranormal|Vampire...   

                                           image_url  
0  https://images.gr-assets.com/books/1447303603l...  
1  https://images.gr-assets.com/books/1255614970l...  
2  https://images.gr-assets.com/books/1361975680l...  
3  https://images.gr-assets.com/books/1320399351l...  
4  https://images.gr-assets.com/books/1361039443l...

<div style="background-color:cornsilk;"><span style="color:#182B49;">Checking for null cells in the dataset </span></div>

In [7]:
#columns which contain null values and the number of null elements
null_counts = books.isnull().sum()
null_counts[null_counts > 0].sort_values(ascending=False) #null_counts in each column (sorted)

book_edition    48848
book_isbn       12866
genres           3242
book_pages       2522
book_format      1656
book_desc        1331
image_url         683
dtype: int64

<span style="color:darkred; font-weight:bold; font-size:13px;">We need to think about how to deal with the null values<br></span>

<div style="background-color:cornsilk;"><span style="color:#182B49;">Exploring dataset columns</span></div>

In [8]:
books.columns 
#we can look over all the current jupyter notebooks in kaggle and pick all the interesting things (including plots, analysis, etc), they have done.

Index(['book_authors', 'book_desc', 'book_edition', 'book_format', 'book_isbn',
       'book_pages', 'book_rating', 'book_rating_count', 'book_review_count',
       'book_title', 'genres', 'image_url'],
      dtype='object')

In [9]:
books['book_authors'] #column "book_authors"

0                                          Suzanne Collins
1                               J.K. Rowling|Mary GrandPré
2                                               Harper Lee
3        Jane Austen|Anna Quindlen|Mrs. Oliphant|George...
4                                          Stephenie Meyer
                               ...                        
54296                                        Howard Megdal
54297                                        Howard Megdal
54298                                        Howard Megdal
54299                               Mimi Baird|Eve Claxton
54300                                           Leah Price
Name: book_authors, Length: 54301, dtype: object

<span style="background-color:cornsilk;">Unique values and range for each column within the 54301 rows</span>

In [10]:
print("Number of unique book_authors: ", len(np.unique(books['book_authors']))) #unique elements in the column
print("Not unique: ", books.shape[0]- len(np.unique(books['book_authors']))) #not unique elements in this column
print("book_authors", np.unique(books['book_authors'])) #unique elements in this column

Number of unique book_authors:  27159
Not unique:  27142
book_authors ['(US)John G. Neihardt' '50 Cent|Robert Greene' "A'Mera Frieman" ...
 '天野 しゅにんた|Amano Shuninta' '籽月' '蘭陵笑笑生|劉本棟|繆天華']


In [11]:
books.head()

book_authors  \
0                                    Suzanne Collins   
1                         J.K. Rowling|Mary GrandPré   
2                                         Harper Lee   
3  Jane Austen|Anna Quindlen|Mrs. Oliphant|George...   
4                                    Stephenie Meyer   

                                           book_desc  \
0  Winning will make you famous. Losing means cer...   
1  There is a door at the end of a silent corrido...   
2  The unforgettable novel of a childhood in a sl...   
3  «È cosa ormai risaputa che a uno scapolo in po...   
4  About three things I was absolutely positive.F...   

                         book_edition book_format    book_isbn book_pages  \
0                                 NaN   Hardcover  9.78044E+12  374 pages   
1                          US Edition   Paperback  9.78044E+12  870 pages   
2                    50th Anniversary   Paperback  9.78006E+12  324 pages   
3  Modern Library Classics, USA / CAN   Paperback  9.78068E+12  279 pages   
4                                 NaN   Paperback  9.78032E+12  498 pages   

   book_rating  book_rating_count  book_review_count  \
0         4.33            5519135             160706   
1         4.48            2041594              33264   
2         4.27            3745197              79450   
3         4.25            2453620              54322   
4         3.58            4281268              97991   

                                  book_title  \
0                           The Hunger Games   
1  Harry Potter and the Order of the Phoenix   
2                      To Kill a Mockingbird   
3                        Pride and Prejudice   
4                                   Twilight   

                                              genres  \
0  Young Adult|Fiction|Science Fiction|Dystopia|F...   
1                        Fantasy|Young Adult|Fiction   
2  Classics|Fiction|Historical|Historical Fiction...   
3                           Classics|Fiction|Romance   
4  Young Adult|Fantasy|Romance|Paranormal|Vampire...   

                                           image_url  
0  https://images.gr-assets.com/books/1447303603l...  
1  https://images.gr-assets.com/books/1255614970l...  
2  https://images.gr-assets.com/books/1361975680l...  
3  https://images.gr-assets.com/books/1320399351l...  
4  https://images.gr-assets.com/books/1361039443l...

<div style="background-color:cornsilk;"><span style="color:#182B49;">Splitting the genres column of the above dataset and
expanding it into multiple genres</span></div>

In [16]:
g= []
for i in range(books.shape[0]):
    if isinstance(books['genres'].iloc[i], str) and len(books['genres'].iloc[i])>0:
        g.append(books['genres'].iloc[i])

In [19]:
#book_genres = books['genres'].str.split('|', expand=False) or above

In [22]:
movie_genres[:5] #Genres of top 5 rows

0    [Young Adult, Fiction, Science Fiction, Dystop...
1                      [Fantasy, Young Adult, Fiction]
2    [Classics, Fiction, Historical, Historical Fic...
3                         [Classics, Fiction, Romance]
4    [Young Adult, Fantasy, Romance, Paranormal, Va...
Name: genres, dtype: object

In [24]:
#figuring out all the existing genres and saving the genres in Genres_list list
Genres_list = []
for i in range(len(books)):
    row = books.iloc[i]
    Gen = str(row.genres)
    genres = Gen.split('|')
    for genre in genres:
        if genre not in Genres_list:
            Genres_list.append(genre)
#len(Genres_list)

In [25]:
len(Genres_list)

867

<div style="background-color:cornsilk;"><span style="color:#182B49;">Adding genres columns into the dataset</span></div>

In [27]:
for genre in Genres_list:
    books[genre] = 0 #Initiating the genres with 0, then each movie's genres will be 1
#df.head()
books.columns

Index(['book_authors', 'book_desc', 'book_edition', 'book_format', 'book_isbn',
       'book_pages', 'book_rating', 'book_rating_count', 'book_review_count',
       'book_title',
       ...
       'Subways', 'London Underground', 'Apple', 'Local History',
       'Disability Studies', 'Peak Oil', 'Social Change', 'Health Care',
       'Read For College', 'Anthropomorphic'],
      dtype='object', length=879)

In [36]:
books.shape

(54301, 879)

<div style="background-color:cornsilk;"><span style="color:#182B49;">Filling the genre columns via 1 if a movie is contained in each genre</span></div>

<span style="color:darkred; font-weight:bold; font-size:13px;">It might be a bad idea to add all genres as columns and we might need to do a better act of grouping some genres together. Now we can treat all genres columns as bags of words. So in the future we can do comparison between movies based on their genres.<br></span>

<div style="background-color:cornsilk;"><span style="color:#182B49;">Merging this dataset with others</span></div>

In [36]:
goodreads = pd.read_csv('books.csv',error_bad_lines = False)

b'Skipping line 3350: expected 12 fields, saw 13\nSkipping line 4704: expected 12 fields, saw 13\nSkipping line 5879: expected 12 fields, saw 13\nSkipping line 8981: expected 12 fields, saw 13\n'


In [42]:
books.columns = ['book_authors', 'book_desc', 'book_edition', 'book_format', 'book_isbn',
       'book_pages', 'book_rating', 'book_rating_count', 'book_review_count',
       'title', 'genres', 'image_url']

In [44]:
goodreads.head()

bookID                                              title  \
0       1  Harry Potter and the Half-Blood Prince (Harry ...   
1       2  Harry Potter and the Order of the Phoenix (Har...   
2       4  Harry Potter and the Chamber of Secrets (Harry...   
3       5  Harry Potter and the Prisoner of Azkaban (Harr...   
4       8  Harry Potter Boxed Set  Books 1-5 (Harry Potte...   

                      authors  average_rating        isbn         isbn13  \
0  J.K. Rowling/Mary GrandPré            4.57  0439785960  9780439785969   
1  J.K. Rowling/Mary GrandPré            4.49  0439358078  9780439358071   
2                J.K. Rowling            4.42  0439554896  9780439554893   
3  J.K. Rowling/Mary GrandPré            4.56  043965548X  9780439655484   
4  J.K. Rowling/Mary GrandPré            4.78  0439682584  9780439682589   

  language_code    num_pages  ratings_count  text_reviews_count  \
0           eng          652        2095690               27591   
1           eng          870        2153167               29221   
2           eng          352           6333                 244   
3           eng          435        2339585               36325   
4           eng         2690          41428                 164   

  publication_date        publisher  
0        9/16/2006  Scholastic Inc.  
1         9/1/2004  Scholastic Inc.  
2        11/1/2003       Scholastic  
3         5/1/2004  Scholastic Inc.  
4        9/13/2004       Scholastic

In [45]:
books.head()

book_authors  \
0                                    Suzanne Collins   
1                         J.K. Rowling|Mary GrandPré   
2                                         Harper Lee   
3  Jane Austen|Anna Quindlen|Mrs. Oliphant|George...   
4                                    Stephenie Meyer   

                                           book_desc  \
0  Winning will make you famous. Losing means cer...   
1  There is a door at the end of a silent corrido...   
2  The unforgettable novel of a childhood in a sl...   
3  «È cosa ormai risaputa che a uno scapolo in po...   
4  About three things I was absolutely positive.F...   

                         book_edition book_format    book_isbn book_pages  \
0                                 NaN   Hardcover  9.78044E+12  374 pages   
1                          US Edition   Paperback  9.78044E+12  870 pages   
2                    50th Anniversary   Paperback  9.78006E+12  324 pages   
3  Modern Library Classics, USA / CAN   Paperback  9.78068E+12  279 pages   
4                                 NaN   Paperback  9.78032E+12  498 pages   

   book_rating  book_rating_count  book_review_count  \
0         4.33            5519135             160706   
1         4.48            2041594              33264   
2         4.27            3745197              79450   
3         4.25            2453620              54322   
4         3.58            4281268              97991   

                                       title  \
0                           The Hunger Games   
1  Harry Potter and the Order of the Phoenix   
2                      To Kill a Mockingbird   
3                        Pride and Prejudice   
4                                   Twilight   

                                              genres  \
0  Young Adult|Fiction|Science Fiction|Dystopia|F...   
1                        Fantasy|Young Adult|Fiction   
2  Classics|Fiction|Historical|Historical Fiction...   
3                           Classics|Fiction|Romance   
4  Young Adult|Fantasy|Romance|Paranormal|Vampire...   

                                           image_url  
0  https://images.gr-assets.com/books/1447303603l...  
1  https://images.gr-assets.com/books/1255614970l...  
2  https://images.gr-assets.com/books/1361975680l...  
3  https://images.gr-assets.com/books/1320399351l...  
4  https://images.gr-assets.com/books/1361039443l...

<div style="background-color:cornsilk;"><span style="color:#182B49;">Converting values in column title to lower (for a better merge)</span></div>

In [52]:
books['title'].iloc[0].lower()

'the hunger games'

In [54]:
lowered = [books['title'].iloc[i].lower() for i in range(len(books['title']))]
books['title'] = lowered

In [55]:
goodreads['title'].iloc[0].lower()

'harry potter and the half-blood prince (harry potter  #6)'

In [56]:
#future: we need to get rid of the "(", ",", etc in titles for a betterr merge

In [66]:
lowered = [goodreads['title'].iloc[i].lower() for i in range(len(goodreads['title']))]
goodreads['title'] = lowered

In [70]:
merged = pd.merge(books, goodreads, on='title')
merged

book_authors  \
0                       Harper Lee   
1                       Harper Lee   
2                       Harper Lee   
3                       Harper Lee   
4                       Harper Lee   
...                            ...   
5416                Kazuo Ishiguro   
5417                   Martin Amis   
5418               Joseph Wambaugh   
5419  J.G. Ballard|Anthony Burgess   
5420              Garrison Keillor   

                                              book_desc      book_edition  \
0     The unforgettable novel of a childhood in a sl...  50th Anniversary   
1     The unforgettable novel of a childhood in a sl...  50th Anniversary   
2     The unforgettable novel of a childhood in a sl...  50th Anniversary   
3     Harper Lee's Pulitzer Prize-winning masterwork...               NaN   
4     Harper Lee's Pulitzer Prize-winning masterwork...               NaN   
...                                                 ...               ...   
5416  From the Booker Prize-winning, bestselling aut...               NaN   
5417  In his uproarious first novel Martin Amis, aut...               NaN   
5418  A cheap hooker named Missy Moonbeam takes a fa...               NaN   
5419  First published in 1978, this collection of ni...               NaN   
5420  In this thoughtful, deeply personal work, one ...               NaN   

                book_format    book_isbn book_pages  book_rating  \
0                 Paperback  9.78006E+12  324 pages         4.27   
1                 Paperback  9.78006E+12  324 pages         4.27   
2                 Paperback  9.78006E+12  324 pages         4.27   
3     Mass Market Paperback   9.7801E+12  309 pages         4.27   
4     Mass Market Paperback   9.7801E+12  309 pages         4.27   
...                     ...          ...        ...          ...   
5416              Paperback  9.78038E+12  336 pages         3.47   
5417              Paperback  9.78068E+12  240 pages         3.59   
5418              Paperback  9.78055E+12  291 pages         3.63   
5419              Paperback  9.78031E+12  302 pages         4.20   
5420              Paperback  9.78014E+12  288 pages         3.97   

      book_rating_count  book_review_count  \
0               3745197              79450   
1               3745197              79450   
2               3745197              79450   
3               3746569              79475   
4               3746569              79475   
...                 ...                ...   
5416              21291               2037   
5417               8404                397   
5418                755                 22   
5419               1441                 92   
5420               1266                154   

                                                  title  ...  \
0                                 to kill a mockingbird  ...   
1                                 to kill a mockingbird  ...   
2                                 to kill a mockingbird  ...   
3                                 to kill a mockingbird  ...   
4                                 to kill a mockingbird  ...   
...                                                 ...  ...   
5416                               when we were orphans  ...   
5417                                  the rachel papers  ...   
5418                                     the delta star  ...   
5419                             the best short stories  ...   
5420  homegrown democrat: a few plain thoughts from ...  ...   

                                  authors average_rating        isbn  \
0                              Harper Lee           4.27  0060935464   
1                 Harper Lee/Sissy Spacek           4.27  0060888695   
2                              Harper Lee           4.27  0061205699   
3                              Harper Lee           4.27  0060935464   
4                 Harper Lee/Sissy Spacek           4.27  0060888695   
...                                   ...            ...         ... 

In [71]:
goodreads.shape, books.shape, merged.shape

((11123, 12), (54301, 12), (5421, 23))